### PR TITLE
feat(headless): implement headless Menu

### LINKED ISSUE #1518: Fix OnyxUserMenu captures click events before they reach the actual target
### Version

1.0.0-beta.0

### Description

Hi, while trying to use the OnyxUserMenu I encountered the following bug:

Whenever I try to click something inside the menu it just closes, blocking any interaction with the elements inside of it.
Using the keyboard to interact with the element works as expected but click events seem to be blocked at some point.

### Reproduction steps

Here's a playground reproducing the issue. The example is basically the same as in the [docs](https://onyx.schwarz/development/theming.html#dark-mode), just without vueuse.

[Playground](https://playground.onyx.schwarz/#eNp9Uk1PAjEQ/SuTekATPg56QjAB5ACJYERvTUyzDLDQbZu2u2II/91pF9g1IfTUznsz895MD2xgTLvIkXVZzyU2NR4c+tyAFGrd58w7zl64SjOjrYcDzNXvfiaKobDNeP9yaN9Q5eVrpKW2i2SDGYbgxGMGR1hZnQFnLvUtTSTOnmsFLa4qCgmJKFeJVs5DUtWDfqDeN5bC7hoPxOl1Sr2kjh7UyUjhkV4AvUolCGNaSmRIXsZ7QSQMoWgK6FwS4Y5aetz7gUVxAk+Vzh4hp8up1lRvFLxqknvhntjXZlC0Mr1ESXk1R5xBp2rUqXc6i+v8txU5pS8K1FDWpEWR/lW6bm+dVrTNQ8gI/TKTSrRz41MaKWddiEjAhJT6Zxpj3ubYPMdJXbK7Et862l2XLu8WSWhBDi6YF3aNvoTHixnNsQaS+VwS+wb4gU7LPGgsacNcLUl2jRfVTuK3SdX60433HpU7mwpCA/MY+fEnjW5Yr+Q+tp9iHldHmuJ3gTbUpAGGwTrPjn95Lg0B)

### Confirmations

- [X] I have checked that there is not already a bug report for my issue.
- [X] I confirm that are no sensitive information included in this bug report. e.g. credentials, internal project details (for employees of the Schwarz Group) etc.

### PR DESCRIPTION
**Pull request by JoCa96**
Relates to #1415, #883
Closes #1518 

- implemented headless feature: `createNavigationMenu`
- headless MenuButton:
  -  now takes an `isExpandedRef` and `onToggle` via it's options
  - `flyout` element is removed as it is not needed
  - removed hover and focus toggle features and moved them to the onyx component directly as these are non spec features

### CODE REVIEW

File: packages/headless/src/composables/menuButton/createMenuButton.ts [hunk 1]
```diff
@@ -1,125 +1,114 @@
-import { computed, ref } from "vue";
+import { computed, unref, type MaybeRef } from "vue";
 import { createBuilder } from "../../utils/builder";
 import { createId } from "../../utils/id";
-import { debounce } from "../../utils/timer";
+
+type CreateMenuButtonOptions = {
+  isExpanded: MaybeRef<boolean>;
```
Comment (larsrickert) [inline packages/headless/src/composables/menuButton/createMenuButton.ts:6]:
```suggestion
  isExpanded: Ref<boolean>;
```

It does not really make sense that its always opened or closed, right?

File: packages/headless/src/composables/menuButton/createMenuButton.ts [hunk 2]
```diff
@@ -1,125 +1,114 @@
...
-    if (currentMenuItem) {
-      const currentIndex = menuItems.indexOf(currentMenuItem);
-      switch (next) {
-        case "next":
-          nextIndex = currentIndex + 1;
+    const handleKeydown = (event: KeyboardEvent) => {
+      switch (event.key) {
+        case "ArrowDown":
+        case "ArrowRight":
+          event.preventDefault();
```
Comment (larsrickert) [inline packages/headless/src/composables/menuButton/createMenuButton.ts:57]: If we preventDefault for all cases should we move it above the switch?

File: packages/sit-onyx/src/components/OnyxFlyoutMenu/future/OnyxFlyoutMenu.vue [hunk 3]
```diff
@@ -24,23 +39,39 @@ const slots = defineSlots<{
...
 
 <template>
-  <div class="onyx-future-flyout-menu">
+  <!-- eslint-disable-next-line vuejs-accessibility/no-static-element-interactions -->
+  <div
+    class="onyx-future-flyout-menu"
+    @focusin="open = true"
+    @focusout="open = false"
+    @mouseover="() => updateDebouncedExpanded(true)"
+    @mouseout="() => updateDebouncedExpanded(false)"
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxFlyoutMenu/future/OnyxFlyoutMenu.vue:69]: Why did you move this out of the composable?

File: packages/sit-onyx/src/components/OnyxUserMenu/types.ts [hunk 4]
```diff
@@ -13,4 +13,8 @@ export type OnyxUserMenuProps = {
    * Optional user description that is displayed when the menu is opened.
    */
   description?: string;
+  /**
+   * If the flyout is expanded or not.
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxUserMenu/types.ts:17]:
```suggestion
   * If the flyout is expanded or not.
   If `undefined`, the state will be managed internally.
```

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/headless/src/composables/menuButton/createMenuButton.ts (hunk 2 above)
Comment (JoCa96) [reply packages/headless/src/composables/menuButton/createMenuButton.ts:57]: No, then it will also prevent the default of all other key presses

File: packages/sit-onyx/src/components/OnyxFlyoutMenu/future/OnyxFlyoutMenu.vue (hunk 3 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxFlyoutMenu/future/OnyxFlyoutMenu.vue:69]: see PR description: These are features that are not in the pattern or spec.

File: packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxUserMenu/types.ts [hunk 5]
```diff
@@ -11,4 +11,9 @@ export type OnyxUserMenuProps = {
    * Optional user description that is displayed when the menu is opened.
    */
   description?: string;
+  /**
+   * If the flyout is expanded or not.
+   * If `undefined`, the state will be managed internally.
+   */
+  open?: boolean;
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxUserMenu/types.ts:18]: I don't see this property used anywhere?

File: packages/headless/src/composables/navigationMenu/createMenu.testing.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+   * Locator for the button elements.
+   */
+  buttons: Locator;
+};
+
+/**
+ * Playwright utility for executing accessibility testing for a navigation menu.
+ * Will check aria attributes and keyboard shortcuts as defined in https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation/
+ */
+export const navigationTesting = async ({ nav, buttons }: NavigationMenuTestingOptions) => {
```
Comment (larsrickert) [inline packages/headless/src/composables/navigationMenu/createMenu.testing.ts:19]: Seems like this is not used in any onyx component 🤔

File: packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxMenuItem/OnyxMenuItem.vue [hunk 7]
```diff
@@ -14,7 +14,7 @@ const emit = defineEmits<{
 
 const {
   elements: { listItem, menuItem },
-} = createMenuItem();
+} = createMenuItems();
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxMenuItem/OnyxMenuItem.vue:17]: Regarding the plural name of the composable: Isn't this only 1 menu item?

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/headless/src/composables/navigationMenu/createMenu.testing.ts (hunk 6 above)
Comment (JoCa96) [reply packages/headless/src/composables/navigationMenu/createMenu.testing.ts:19]: Will do in follow-up

File: packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxUserMenu/types.ts (hunk 5 above)
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxNavBar/modules/OnyxUserMenu/types.ts:18]: Will do in follow-up

**Review by larsrickert — APPROVED** [review]
